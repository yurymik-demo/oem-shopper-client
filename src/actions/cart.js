import { reduce } from "lodash";
import {
  GET_USER_CART,
  CART_ADD_ITEM,
  UPDATE_CART_TOTALS
} from "./types.const";

// Mock cart data until connected to server API
const _MOCK_USER_CART = {
  // will have a nested 'item', id and a quanitity fields
  cartItems: [],
  totals: {
    itemsCount: 0
  }
};

const calculateCartTotals = cartItems => {
  const itemsCount = reduce(
    cartItems || [],
    (sum, cartItem) => sum + cartItem.quantity,
    0
  );

  return {
    itemsCount
  };
};

export const getUserCart = () => dispatch => {
  // adding timeout to simulate the api call and to demonstrate "waiting" experience
  setTimeout(() => {
    dispatch({ type: GET_USER_CART, data: _MOCK_USER_CART });
  }, 1000);
};

export const addToCart = (item, quantity = 1) => dispatch => {
  // MOCK storage for user cart items
  // when have real API this will be replaced with an axios call

  // check if item is already in the cart
  const existingItem = _MOCK_USER_CART.cartItems.find(
    cartItem => cartItem.id === item.id
  );

  // if new items, add it to cartItems collection, if existing, increase quantity
  if (!existingItem) {
    _MOCK_USER_CART.cartItems.push({ id: item.id, item, quantity });
  } else {
    existingItem.quantity += quantity;
  }

  _MOCK_USER_CART.totals = calculateCartTotals(_MOCK_USER_CART.cartItems);

  dispatch({
    type: CART_ADD_ITEM,
    cartItems: _MOCK_USER_CART.cartItems,
    totals: _MOCK_USER_CART.totals
  });
};

export const updateCartTotals = () => dispatch => {
  _MOCK_USER_CART.totals = calculateCartTotals(_MOCK_USER_CART.cartItems);
  dispatch({ type: UPDATE_CART_TOTALS, totals: _MOCK_USER_CART.totals });
};

export const removeItemFromCart = itemId => dispatch => {
  const newItemsCollection = _MOCK_USER_CART.cartItems.filter(
    item => item.id !== itemId
  );
  _MOCK_USER_CART.cartItems = newItemsCollection;
  dispatch({ type: GET_USER_CART, data: _MOCK_USER_CART });
  updateCartTotals();
};
