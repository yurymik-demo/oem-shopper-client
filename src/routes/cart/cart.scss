.cart-container {
  .action-button {
    margin-left: 8px;
  }
}
