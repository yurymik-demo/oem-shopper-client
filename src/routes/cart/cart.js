import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import ProgressIndicator from "../../components/ProgressIndicator/ProgressIndicator";
import { getUserCart } from "../../actions/cart";
import CartItemCard from "../../components/CartItemCard/CartItemCard";

import "./cart.scss";

export class Cart extends React.Component {
  componentDidMount() {
    this.props.getUserCart();
  }

  renderNoItemsView = () => {
    return <h4>Your Cart is empty.</h4>;
  };

  renderCartItems = items => {
    if (!items) return <ProgressIndicator />;
    if (!items.length) return this.renderNoItemsView();
    return items.map(cartItem => {
      return (
        <div key={cartItem.id}>
          <CartItemCard key={cartItem.id} cartItem={cartItem} />
        </div>
      );
    });
  };

  render() {
    const { cartItems: items } = this.props.userCart || {};
    const canCheckout = items && items.length > 0;
    return (
      <div className="cart-container">
        <h1>Your Cart</h1>
        {this.renderCartItems(items)}
        <div className="row">
          {canCheckout && (
            <Link
              to="/checkout"
              className="action-button btn-flat orange right"
            >
              Checkout
            </Link>
          )}
          <Link to="/" className="action-button btn-flat orange right">
            Continue Chopping
          </Link>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ userCart }) => ({
  userCart
});

export default connect(
  mapStateToProps,
  { getUserCart }
)(Cart);
