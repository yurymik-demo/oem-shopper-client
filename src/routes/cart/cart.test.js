import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import { Cart } from "./cart";

jest.mock("../../components/ItemsList/ItemsList", () => "ItemsListComponent");

it("renders without crashing", () => {
  const getUserCartMock = jest.fn(() => {});
  global.scrollTo = jest.fn(() => {});
  const wrapper = renderer
    .create(
      <BrowserRouter>
        <Cart getUserCart={getUserCartMock} />
      </BrowserRouter>
    )
    .toJSON();
  expect(wrapper).toMatchSnapshot();
  expect(getUserCartMock).toHaveBeenCalled();
});
