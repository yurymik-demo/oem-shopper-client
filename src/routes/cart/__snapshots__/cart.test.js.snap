// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="cart-container"
>
  <h1>
    Your Cart
  </h1>
  <div
    className="progress"
  >
    <div
      className="indeterminate"
    />
  </div>
  <div
    className="row"
  >
    <hr
      className="col s12"
    />
  </div>
  <div
    className="row"
  >
    <a
      className="action-button btn-flat orange right"
      href="/"
      onClick={[Function]}
    >
      Continue Shopping
    </a>
  </div>
</div>
`;
